File: src/scripts/parse_receipt.py
import os
import sys
import json
import base64
from dotenv import load_dotenv
load_dotenv()

import anthropic

from prompts import DEFAULT_RECEIPT_PARSING_PROMPT

VALID_CATEGORIES = ['meals', 'travel', 'lodging', 'software', 'equipment', 'other']


def parse_receipt(image_path: str, prompt: str | None = None) -> dict:
    with open(image_path, 'rb') as f:
        image_bytes = f.read()

    ext = image_path.rsplit('.', 1)[-1].lower()
    media_type_map = {
        'jpg': 'image/jpeg',
        'jpeg': 'image/jpeg',
        'png': 'image/png',
        'gif': 'image/gif',
        'webp': 'image/webp',
    }
    media_type = media_type_map.get(ext, 'image/jpeg')
    image_b64 = base64.standard_b64encode(image_bytes).decode('utf-8')

    client = anthropic.Anthropic()

    prompt = prompt or DEFAULT_RECEIPT_PARSING_PROMPT

    response = client.messages.create(
        model="claude-opus-4-8",
        max_tokens=512,
        thinking={"type": "adaptive"},
        messages=[
            {
                "role": "user",
                "content": [
                    {
                        "type": "image",
                        "source": {
                            "type": "base64",
                            "media_type": media_type,
                            "data": image_b64,
                        },
                    },
                    {"type": "text", "text": prompt},
                ],
            }
        ],
    )

    response_text = next(
        (block.text for block in response.content if block.type == "text"), "{}"
    )

    try:
        data = json.loads(response_text)
    except json.JSONDecodeError:
        data = {
            "amount": 0.0,
            "category": "other",
            "vendor": "",
            "description": "",
            "chargeToClient": False,
            "accuracy": 0.0,
        }

    # Normalise category
    if data.get("category") not in VALID_CATEGORIES:
        data["category"] = "other"

    data["amount"] = float(data.get("amount") or 0)
    data["accuracy"] = float(data.get("accuracy") or 0)
    data["chargeToClient"] = bool(data.get("chargeToClient", False))

    return data


if __name__ == "__main__":
    payload = json.loads(sys.stdin.read())
    result = parse_receipt(payload["image_path"])
    print(json.dumps(result))

File: src/lib/parseReceipt.ts
import Anthropic from '@anthropic-ai/sdk'
import { DEFAULT_RECEIPT_PARSING_PROMPT } from './prompts'

const VALID_CATEGORIES = ['meals', 'travel', 'lodging', 'software', 'equipment', 'other'] as const

export interface ParsedReceipt {
  amount: number
  category: string
  vendor: string
  description: string
  chargeToClient: boolean
  accuracy: number
}

export async function parseReceipt(
  imageBuffer: Buffer,
  mediaType: 'image/jpeg' | 'image/png' | 'image/gif' | 'image/webp',
  prompt?: string | null
): Promise<ParsedReceipt> {
  const imageB64 = imageBuffer.toString('base64')
  const client = new Anthropic()

  const response = await client.messages.create({
    model: 'claude-opus-4-8',
    max_tokens: 512,
    messages: [
      {
        role: 'user',
        content: [
          {
            type: 'image',
            source: {
              type: 'base64',
              media_type: mediaType,
              data: imageB64,
            },
          },
          { type: 'text', text: prompt ?? DEFAULT_RECEIPT_PARSING_PROMPT },
        ],
      },
    ],
  })

  const textBlock = response.content.find((b) => b.type === 'text')
  const text = textBlock && 'text' in textBlock ? textBlock.text : '{}'

  let data: Record<string, any>
  try {
    data = JSON.parse(text)
  } catch {
    data = { amount: 0, category: 'other', vendor: '', description: '', chargeToClient: false, accuracy: 0 }
  }

  if (!VALID_CATEGORIES.includes(data.category)) data.category = 'other'

  return {
    amount: parseFloat(data.amount ?? 0),
    category: data.category,
    vendor: data.vendor ?? '',
    description: data.description ?? '',
    chargeToClient: Boolean(data.chargeToClient),
    accuracy: parseFloat(data.accuracy ?? 0),
  }
}
